core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java

```diff
@@ -125,6 +132,26 @@ public void clusterChanged(ClusterChangedEvent event) {
                     }
                 }
             }
+
+            long nextDelaySetting = UnassignedInfo.findNextDelayedAllocationSetting(settings, event.state());
+            if (nextDelaySetting > 0 && nextDelaySetting < registeredNextDelaySetting) {
+                FutureUtils.cancel(registeredNextDelayFuture);
+                registeredNextDelaySetting = nextDelaySetting;
+                TimeValue nextDelay = TimeValue.timeValueMillis(UnassignedInfo.findNextDelayedAllocationIn(settings, event.state()));
+                logger.info("delaying allocation for [{}] unassigned shards, next check in [{}]", UnassignedInfo.getNumberOfDelayedUnassigned(settings, event.state()), nextDelay);
+                registeredNextDelayFuture = threadPool.schedule(nextDelay, ThreadPool.Names.SAME, new AbstractRunnable() {
+                    @Override
+                    protected void doRun() throws Exception {
+                        routingTableDirty = true;
+                        reroute();
+                    }
+
+                    @Override
+                    public void onFailure(Throwable t) {
+                        logger.warn("failed to schedule/execute reroute post unassigned shard", t);
+                    }
+                });
+            }
```

Comment:
can we add a trace log on as an else so we know why we skipped scheduling if we need to?